Add Immediate uiMode

kenrb opened the pull request:
This specifies the behaviour of the `mediation: 'immediate'`.

Issue: https://github.com/w3c/webauthn/issues/2228
Explainer: https://github.com/w3c/webauthn/wiki/Explainer:-WebAuthn-immediate-mediation

Parallel PR adding the enum to Credential Management: https://github.com/w3c/webappsec-credential-management/pull/272


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/kenrb/webauthn/pull/2291.html" title="Last updated on Aug 5, 2026, 6:09 PM UTC (13cb081)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/webauthn/2291/146d0db...kenrb:13cb081.html" title="Last updated on Aug 5, 2026, 6:09 PM UTC (13cb081)">Diff</a>

emlun committed:
Add JSON partial dictionaries for extensions

kenrb committed:
Add Immediate Mediation mode

This PR adds the Immediate Mediation mode for WebAuthn requests,
which shows discoverable credentials to the user if any are
silently found, or else throws a `NotAllowedError` if there are
none.

See the
[explainer](https://github.com/w3c/webauthn/wiki/Explainer:-WebAuthn-immediate-mediation) for more context.
A parallel PR to add the mode to the Credential Management
specification will be created.

kenrb committed:
Fix link issue

kenrb committed:
Tweaks

kenrb committed:
Added some more bits

kenrb committed:
Merge branch 'w3c:main' into immediate-mediation

Firstyear commented:
Having webauthn prompts randomly trigger on a page load with no explanation or context for users, sounds like the exact opposite of good UX to me - it sounds a lot more like a path to confusion and frustration.

kenrb commented:
> This could still happen after the user navigates to the login page or exactly where it does now. It’s not random we leave it up to RPs - we had this exact same dicussion with WebAuthn user gestures for Safari and they were lifted. As someone responsible for large consumer RP implementations, I have problems seeing clearly how this approach helps for most pages.

The main advantage of not having a user gesture requirement for existing modal WebAuthn calls is that they can be used for re-auth, a use case for which immediate mediation isn't useful.

Immediate is aimed at scenarios in which a user has done something to indicate a sign-in is appropriate at that time. This isn't precisely replicating `preferImmediatelyAvailableCredentials` on mobile because the web has different privacy properties.

There is a separate proposal for a mode called Ambient, in which more subtle (non-modal) UI is displayed to offer the user an opportunity to sign-in, and would not require user activation. https://github.com/w3c/webauthn/wiki/Explainer:-WebAuthn-Ambient-Signin-UI

That proposal is still active.

kenrb committed:
Merge branch 'w3c:main' into immediate-mediation

kenrb committed:
Update index.bs

Added uiMode. Also expanded the privacy section.

kenrb committed:
Merge branch 'w3c:main' into immediate-mediation

kenrb committed:
Update index.bs

kenrb reviewed (commented):
I've updated the PR incorporating @nsatragno's comments

nsatragno reviewed (approved):
Approved, but this shouldn't land until https://github.com/w3c/webappsec-credential-management/pull/272 lands.

kenrb committed:
Update index.bs

Change `CredentialUiModeRequirement` to `CredentialUiMode`, reflecting the same change in the Credential Management PR.

kenrb committed:
Merge branch 'w3c:main' into immediate-mediation

kenrb committed:
Update index.bs

Remove the requirement to consume the transient user activation during an Immediate mode Get request.

kenrb committed:
Update index.bs

Remove requirement for Immediate mode requests to consume user activations (attempt #2).

zacknewman commented on index.bs:
```diff
@@ -2460,7 +2475,7 @@ When this method is invoked, the user agent MUST execute the following algorithm
             and [=set/remove=] |authenticator| from |issuedRequests|. Throw a "{{NotAllowedError}}" {{DOMException}}.
 
-        :   If <code>|options|.{{CredentialRequestOptions/signal}}</code> is present and [=AbortSignal/aborted=],
+        :   If <code>|options|.{{CredentialRequestOptions/signal}}</code> is present and [=AbortSignal/aborted=], and <code>|options|.{{CredentialRequestOptions/mediation}}</code> is not {{CredentialMediationRequirement/immediate}},
```
This is incorrect. This should be `<code>|options|.{{CredentialRequestOptions/uiMode}}</code> is not {{CredentialUiMode/immediate}}`

kenrb committed:
Merge branch 'w3c:main' into immediate-mediation

kenrb committed:
Update index.bs

Replaced some incorrect references to  `CredentialMediationRequirement`.

kenrb committed:
Merge branch 'w3c:main' into immediate-mediation

kenrb committed:
Fix indentation problems

kenrb committed:
Remove errant whitespace

kenrb committed:
Merge branch 'w3c:main' into immediate-mediation
